<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>partition</title>
</head>
<body>
partition
函数签名: partition(predicate: function: boolean, thisArg: any): [Observable, Observable]
Split one observable into two based on provided predicate.
</body>
<script src="https://unpkg.com/rxjs/bundles/rxjs.umd.min.js"></script>
<script type="text/javascript">
    // RxJS v6+
    const { from, merge,of } =rxjs;
    const { partition, map,catchError } =rxjs.operators;

    /**示例 1: 分割偶数和奇数**/
    const source = from([1, 2, 3, 4, 5, 6]);
    // 第一个值(events)返回 true 的数字集合，第二个值(odds)是返回 false 的数字集合
    const [evens, odds] = source.pipe(partition(val => val % 2 === 0));
    /*
      输出:
      "Even: 2"
      "Even: 4"
      "Even: 6"
      "Odd: 1"
      "Odd: 3"
      "Odd: 5"
    */
    const subscribe = merge(
        evens.pipe(map(val => `Even: ${val}`)),
        odds.pipe(map(val => `Odd: ${val}`))
    ).subscribe(val => console.log(val));

    /**示例 2: 分割正常执行和错误**/
    const source2 = from([1, 2, 3, 4, 5, 6]);
    // 如果大于3就抛出错误
    const example2 = source2.pipe(
        map(val => {
            if (val > 3) {
                throw `${val} greater than 3!`;
            }
            return { success: val };
        }),
        catchError(val => of({ error: val }))
    );
    // 分割正常执行或错误
    const [success, error] = example2.pipe(partition(res => res.success));
    /*
      输出:
      "Success! 1"
      "Success! 2"
      "Success! 3"
      "Error! 4 greater than 3!"
    */
    const subscribe2 = merge(
        success.pipe(map(val => `Success! ${val.success}`)),
        error.pipe(map(val => `Error! ${val.error}`))
    ).subscribe(val => console.log(val));

</script>
</html>
